File: [6] ZigZag Conversion/6.zig-zag-conversion.cpp
/*
 * @lc app=leetcode id=6 lang=cpp
 *
 * [6] ZigZag Conversion
 */

// @lc code=start
class Solution {
public:
    string convert(string s, int numRows) {
        if (s.length() < numRows) return s;
        if (numRows <= 1) return s;
        vector<string> zig(numRows);
        bool flag = true;
        int row = 0;
        for(int i = 0; i < s.length(); ++i)
        {
            zig[row].push_back(s[i]);
            if (row == 0) flag = true;
            else if (row == numRows-1) flag = false;
            if (flag) row++;
            else row--;
        }
        string res;
        for(auto& temp: zig)
            res += temp;
        return res;
    }
};
// @lc code=end

File: [6] ZigZag Conversion/6.zig-zag-conversion.java
/*
 * @lc app=leetcode id=6 lang=java
 *
 * [6] ZigZag Conversion
 */

// @lc code=start
class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1) return s;
        if (s.length() <= numRows) return s;
        StringBuilder res = new StringBuilder();
        for (int i = 0; i < numRows; i++) {
            int j = i;
            boolean go_up = true;
            StringBuilder line = new StringBuilder();
            while (j < s.length()) {
                line.append(s.charAt(j));
                if (i == 0 || i == numRows-1) {
                    j += 2*(numRows - 1);
                } else {
                    j += (go_up)? 2*(numRows - 1 - i): 2*i;
                    go_up = !go_up;
                }
            }
            res.append(line);
        }
        return res.toString();
    }
}
// @lc code=end

